bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/RecoveryManager.java

```diff
@@ -0,0 +1,367 @@
+            for (XAResource xaResource : xaResources) {
+                Map<String, Xid> recoveredXids;
+                try {
+                    recoveredXids = retrievePreparedXids(xaResource);
+                } catch (XAException e) {
+                    diagnosticLog.warn(ErrorCodes.TRANSACTION_CANNOT_COLLECT_XIDS_IN_RESOURCE, null, xaResource);
+                    failedToRecoverResources.add(xaResource);
+                    continue;
+                }
+                preparedXids.put(xaResource, recoveredXids);
+            }
+
+            // Check if all the prepared xids are in the pending logs and recover them
+            for (Map.Entry<XAResource, Map<String, Xid>> entry : preparedXids.entrySet()) {
+                allOk = allOk && recoverXAResource(entry.getKey(), entry.getValue());
+            }
+        }
+
+        if (!allOk) {
+            diagnosticLog.error(ErrorCodes.TRANSACTION_STARTUP_RECOVERY_FAILED, null);
+        }
+
+        // notify the user of all startup recovery warns and errors
+        if (!diagnosticLog.getDiagnosticList().isEmpty()) {
+            RuntimeUtils.handleDiagnosticErrors(diagnosticLog);
+        }
+        return allOk;
+    }
+
+    /**
+     * Recovers all the XAResources
+     */
+    public boolean recoverXAResource(XAResource xaResource, Map<String, Xid> recoveredXids) {
+        boolean allXidsRecovered = true;
+        for (Xid xid : recoveredXids.values()) {
+            String globalTransactionIdStr = new String(xid.getGlobalTransactionId());
+            String branchQualifierStr = new String(xid.getBranchQualifier());
+            String combinedIdStr = globalTransactionIdStr + ":" + branchQualifierStr; // combined id
+            if (failedTransactions.containsKey(combinedIdStr)) {
+                RecoveryState state = failedTransactions.get(combinedIdStr).getTransactionState();
+                boolean recovered = recoverFailedTrxInXAResource(xaResource, xid, state);
+                if (recovered) {
+                    // put a terminated log record to indicate that the transaction is recovered successfully
+                    TransactionLogRecord terminatedRecord = new TransactionLogRecord(
+                            globalTransactionIdStr, branchQualifierStr, RecoveryState.TERMINATED);
+                    TransactionResourceManager.getInstance().getLogManager().put(terminatedRecord);
+                    failedTransactions.remove(combinedIdStr);
+                } else {
+                    allXidsRecovered = false;
+                    failedToRecoverResources.add(xaResource);
+                }
+            }
+        }
+        return allXidsRecovered;
+    }
+
+    private boolean recoverFailedTrxInXAResource(XAResource xaResource, Xid xid, RecoveryState state) {
+        boolean recovered = false;
+        switch (state) {
+            case PREPARING: // failed during prepare, no decision record found
+                recovered = handleAbort(xaResource, xid);
```

Comment:
```suggestion
                return handleAbort(xaResource, xid);
```

Reply:
Can apply for other places and return `false` in the default case. wdyt?

Reply:
+1. fixed in [4fd4179](https://github.com/ballerina-platform/ballerina-lang/pull/42080/commits/4fd4179e55c2758cacf679862c6ca4a2988b36d8)